### PR TITLE
Implement chalk unification routines

### LINKED ISSUE #54935: implement the `ResolventOps` methods for chalk integration
The methods in the `ResolventOps` trait are currently unimplemented:

https://github.com/rust-lang/rust/blob/e1643a8968753226dab7ab3c9fee826f097550f2/src/librustc_traits/chalk_context.rs#L275-L297

These methods perform various unification-like operations and will need to be implemented in rustc.

### PR DESCRIPTION
**Pull request by scalexm**
`ResolventOps` and `AggregateOps` are mostly straightforwardly translated from chalk. I had caught a few bugs already in my `chalk` branch and backported fixes to this branch, but there may be other ones left. EDIT: I hope there are none left now :)

Fixes #54935.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_mir/borrow_check/nll/type_check/relate_tys.rs [hunk 1]
```diff
@@ -115,4 +116,16 @@ impl TypeRelatingDelegate<'tcx> for NllTypeRelatingDelegate<'_, '_, '_, 'tcx> {
                 });
         }
     }
+
+    fn push_domain_goal(&mut self, _: DomainGoal<'tcx>) {
+        // No-op
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/nll/type_check/relate_tys.rs]: Maybe `panic!("should never be invoked with eager normalization")`?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Everything seems fine except for the question of unifying two existential inference variables...

File: src/librustc/infer/nll_relate/mod.rs [hunk 2]
```diff
@@ -609,17 +731,89 @@ where
...
+                        }
+                        TypeVariableValue::Unknown { universe } => {
+                            if self.universe.cannot_name(universe) {
+                                debug!(
+                                    "TypeGeneralizer::tys: root universe {:?} cannot name\
+                                    variable in universe {:?}",
+                                    self.universe,
+                                    universe
+                                );
+                                return Err(TypeError::Mismatch);
```
[the remark this answers is not in the document]
Comment (scalexm) [reply src/librustc/infer/nll_relate/mod.rs]: yes that's definitely a mistake

File: src/librustc/infer/nll_relate/mod.rs [hunk 3]
```diff
@@ -11,30 +11,41 @@
...
+//!   in case we know that there are no unbound type inference variables.
+//!   This is the case for NLL, because at NLL time types are fully inferred
+//!   up-to regions.
 //! - This code uses "universes" to handle higher-ranked regions and
 //!   not the leak-check. This is "more correct" than what rustc does
 //!   and we are generally migrating in this direction, but NLL had to
 //!   get there first.
+//!
+//! Also, this code assumes that there are no bound type vars at all, not even
+//! free ones. This is ok because:
```
[the remark this answers is not in the document]
Comment (scalexm) [reply src/librustc/infer/nll_relate/mod.rs]: no "bound type vars", so e.g. no `for<T> ...`

[the remark this answers is not in the document]
Comment (scalexm) [reply src/librustc/infer/nll_relate/mod.rs]: I replaced with "no bound types at all" (I dropped "vars")

**Comment by scalexm** [comment]
@bors p=1
(I'm setting a priority because I'd like this to land quickly, as I'll be less available starting in December, if anybody has an issue with that please contact me)

**Comment by scalexm** [comment]
cc @rust-lang/infra any idea whether the above error is spurious? I can't reproduce locally on a mac.

**Comment by scalexm** [comment]
@bors retry
I see a similar error in https://github.com/rust-lang/rust/pull/55767#issuecomment-441182619

**Comment by scalexm** [comment]
@bors retry

> ---- net::tcp::tests::connect_timeout_unbound stdout ----
28706thread 'net::tcp::tests::connect_timeout_unbound' panicked at 'called `Result::unwrap_err()` on an `Ok` value: TcpStream { addr: V4(127.0.0.1:3856), peer: V4(127.0.0.1:3856), socket: 800 }', src\libcore\result.rs:1009:5

It seems like this PR will never land!

**Comment by scalexm** [comment]
@bors p=0
In case it fails again, I’m giving back the priority to other PRs.

**Comment by scalexm** [comment]
Same failure as in 
https://github.com/rust-lang/rust/pull/55011#issuecomment-443215660

**Pull request merged**